apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.IntegrationTests/HighPerformancePubsub.cs

```diff
@@ -0,0 +1,163 @@
+
+        private async Task RunBulkMessaging(int messageCount, int minMessagesize, int maxMessageSize, int maxMessagesInFlight, int initialNackCount)
+        {
+            var topicId = _fixture.CreateTopicId();
+            var subscriptionId = _fixture.CreateSubscriptionId();
+
+            Console.WriteLine("BulkMessaging test");
+            Console.WriteLine($"{messageCount} messages; of size [{minMessagesize}, {maxMessageSize}]; max in-flight: {maxMessagesInFlight}");
+
+            // Create topic
+            var topicName = new TopicName(_fixture.ProjectId, topicId);
+            var publisher = await PublisherClient.CreateAsync().ConfigureAwait(false);
+            await publisher.CreateTopicAsync(topicName).ConfigureAwait(false);
+
+            // Subscribe to the topic
+            var subscriber = await SubscriberClient.CreateAsync().ConfigureAwait(false);
+            var subscriptionName = new SubscriptionName(_fixture.ProjectId, subscriptionId);
+            await subscriber.CreateSubscriptionAsync(subscriptionName, topicName, null, 60).ConfigureAwait(false);
+
+            // Create high-performance Publisher and Subscriber
+            var bulkPublisher = await HiPerfPublisher.CreateAsync(topicName).ConfigureAwait(false);
+            var bulkSubscriber = await HiPerfSubscriber.CreateAsync(subscriptionName).ConfigureAwait(false);
+
+            Console.WriteLine("Topic, Subscription, High-performance Publisher and Subscriber all created");
+
+            // Subscribe
+            object recvLock = new object();
+            int recvCount = 0; // Count of received messages
+            long recvSum = 0L; // Sum of bytes of received messages
+            Task subTask = bulkSubscriber.StartAsync((msg, ct) =>
+            {
+                var localRecvCount = Interlocked.Increment(ref recvCount);
+                if (localRecvCount <= initialNackCount)
+                {
+                    return Task.FromResult(HiPerfSubscriber.Reply.Nack);
+                }
+                Interlocked.Add(ref recvSum, msg.Data.Sum(x => (long)x));
+                if (localRecvCount >= messageCount + initialNackCount)
+                {
+                    // Test finished, so stop subscriber
+                    Task unused = bulkSubscriber.StopAsync(TimeSpan.FromSeconds(15));
+                }
+                // ACK all messages
+                return Task.FromResult(HiPerfSubscriber.Reply.Ack);
+            });
+
+            // Publish
+            var rnd = new Random(1234);
+            var activePubs = new HashSet<Task>();
+            int sentCount = 0;
+            long sentSum = 0L; // Sum of bytes of sent messages
+
+            // Watchdog to report progress and fail test on deadlock
+            int prevSentCount = -1;
+            int prevRecvCount = -1;
+            CancellationTokenSource watchdogCts = new CancellationTokenSource();
+            Task.Run(async () =>
+            {
+                while (true)
```

Comment:
Possibly `while (!watchdogCts.IsCancellationRequested)` just to make it clearer that this *will* end? (It doesn't much matter whether it ends due to spotting that or the Task.Delay being cancelled.)